Map weights to uniqueified tensor names during model loading.

davidsoergel opened the pull request:
Fixes tensorflow/tfjs#34

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tensorflow/tfjs-layers/117)
<!-- Reviewable:end -->

bileschi commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[src/common.ts, line 17 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QdK5O6Nbggy1T03Fg:-L9QdK5O6Nbggy1T03Fh:b-z7hpt7) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L17)):*
> ```
> const nameMap: Map<string, number> = new Map<string, number>();
> ```
Please add a comment that describes this as a map from tensorName to a count of tensors with the same name.

---

*[src/common.ts, line 105 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9Qdqwu3VX8kTnkvOXo:-L9Qdqwu3VX8kTnkvOXp:b-vrn61k) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L105)):*
> ```
> getNonUniqueScopedTensorName
> ```
If we follow the convention that "unique names always include scope" then this can be reduced to `getScopedTensorName` right?

---

*[src/common.ts, line 105 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QeDl1ClsLyK_WiQE9:-L9QeDl1ClsLyK_WiQEA:b12igbq) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L105)):*
> ```XML
>  * @return Scoped name string.
>  */
> function getNonUniqueScopedTensorName(prefix: string): string {
> ```

Shouldn't the argument here be called something like `tensorName` instead of `prefix`?

---

*[src/common.ts, line 117 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QeSLU8MbL6377vCtK:-L9QeSLU8MbL6377vCtL:bkubrgs) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L117)):*
> ```
> prefix
> ```
As above, this seems more like a tensorName than a prefix, unless I'm misunderstanding something

---

*[src/types.ts, line 72 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QYNIWAjBfFE53Fyn3:-L9QYNIWAjBfFE53Fyn4:b-tzlnz2) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/types.ts#L72)):*
> ```XML
>   readonly id: number;
>   readonly name?: string;
>   readonly originalName?: string;
> ```

Some documentation here about the purpose of `originalName` and its distinction from `name` will be helpful for readers coming to read about SymbolicTensor without the context of issue #34

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-:-L9Qm-ni15VM3rTQykFa:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

bileschi commented on src/models_test.ts:
```diff
@@ -211,13 +207,53 @@ describeMathCPU('loadModel', () => {
+      await modelFromJSON(
+          {modelTopology: configJson, weightsManifest, pathPrefix: '.'});
+
+      // On the second load, the variable names will be uniqueified.  This test
```
if a user has a weight named weight_1, and two weights named 'weight' wont there still be a collision?

davidsoergel commented:
Review status: 0 of 5 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[src/common.ts, line 17 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QdK5O6Nbggy1T03Fg:-L9RKmjz2_L7IBenVHEP:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L17)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> const nameMap: Map<string, number> = new Map<string, number>();
> ```
Please add a comment that describes this as a map from tensorName to a count of tensors with the same name.
</blockquote></details>

Done.

---

*[src/common.ts, line 105 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9Qdqwu3VX8kTnkvOXo:-L9RL-aK-Rg202Y4g04S:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L105)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> getNonUniqueScopedTensorName
> ```
If we follow the convention that "unique names always include scope" then this can be reduced to `getScopedTensorName` right?

</blockquote></details>

Done.

---

*[src/common.ts, line 105 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QeDl1ClsLyK_WiQE9:-L9RL8lm4qWi4-nXt6R5:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L105)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

Shouldn't the argument here be called something like `tensorName` instead of `prefix`?
</blockquote></details>

Done.

---

*[src/common.ts, line 117 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QeSLU8MbL6377vCtK:-L9RMjOb1wde4QdyhlpT:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/common.ts#L117)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

> ```
> prefix
> ```
As above, this seems more like a tensorName than a prefix, unless I'm misunderstanding something
</blockquote></details>

Done.

---

*[src/common.ts, line 115 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QW537iTM1XDMxNNdd-r2-115:-L9RMjor4m2Fsh4ccieX:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/b3ef63bcc206027931cc25b3b7fc82955308578c/src/common.ts#L115)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

This is no longer accurate. Update this.
</blockquote></details>

Done.

---

*[src/models_test.ts, line 215 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QW537iTM1XDMxNNdf-r2-215:-L9RN8Fe5KBiZwKBt1Wu:b-t5wg68) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/b3ef63bcc206027931cc25b3b7fc82955308578c/src/models_test.ts#L215)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

nit: No need to use backticks, just use regular quotes.
</blockquote></details>

Done throughout this file

---

*[src/models_test.ts, line 246 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QW537iTM1XDMxNNdf-r2-246:-L9RUho00Y8bJCSIY6_s:b-l88amy) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/b3ef63bcc206027931cc25b3b7fc82955308578c/src/models_test.ts#L246)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

if a user has a weight named weight_1, and two weights named 'weight' wont there still be a collision?
</blockquote></details>

Added asserts showing that the names are uniqueified and that the weights are actually loaded.

Shanqing, you're right that the name-uniqueifying code already had a bug that could produce collisions.  I added a test to exercise that, and fixed it.

---

*[src/models_test.ts, line 254 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QW537iTM1XDMxNNdf-r2-254:-L9RNZ-mD5-uvYSw2tVd:b74j6ke) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/b3ef63bcc206027931cc25b3b7fc82955308578c/src/models_test.ts#L254)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

Optional: I find `e.stack` to be more informative for debugging.
</blockquote></details>

Aha, thanks!  Since Jasmine explicitly accepts an Error here, I thought it should be Jasmine's job to render the stack trace properly, but I couldn't get it to work before.  Fixed.

---

*[src/types.ts, line 72 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QYNIWAjBfFE53Fyn3:-L9RRsj-6Y9X8-o21NlH:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/a2a550f2c1a25b4a7fc4169f21be78c7dee589ca/src/types.ts#L72)):*
<details><summary><i>Previously, bileschi (Stanley Bileschi) wrote…</i></summary><blockquote>

Some documentation here about the purpose of `originalName` and its distinction from `name` will be helpful for readers coming to read about SymbolicTensor without the context of issue #34
</blockquote></details>

Done.

---

*[src/types.ts, line 72 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117#-L9QW537iTM1XDMxNNdg-r2-72:-L9RRtPGAKI9b9ttvbBU:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/b3ef63bcc206027931cc25b3b7fc82955308578c/src/types.ts#L72)):*
<details><summary><i>Previously, caisq (Shanqing Cai) wrote…</i></summary><blockquote>

We need a comment here explaining the difference between `name` and `originalName`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-layers/117)*
<!-- Sent from Reviewable.io -->
